materialize-elasticsearch/type_mapping.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
+
+	pf "github.com/estuary/flow/go/protocols/flow"
+)
+
+type elasticPropertyType string
+
+const (
+	elasticTypeKeyword   elasticPropertyType = "keyword"
+	elasticTypeBoolean   elasticPropertyType = "boolean"
+	elasticTypeLong      elasticPropertyType = "long"
+	elasticTypeDouble    elasticPropertyType = "double"
+	elasticTypeText      elasticPropertyType = "text"
+	elasticTypeBinary    elasticPropertyType = "binary"
+	elasticTypeDate      elasticPropertyType = "date"
+	elasticTypeIp        elasticPropertyType = "ip"
+	elasticTypeFlattened elasticPropertyType = "flattened"
+)
+
+type property struct {
+	Type   elasticPropertyType `json:"type"`
+	Coerce bool                `json:"coerce,omitempty"`
+}
+
+func propForField(field string, binding *pf.MaterializationSpec_Binding) property {
+	return propForProjection(binding.Collection.GetProjection(field))
+}
+
+func propForProjection(p *pf.Projection) property {
+	if t, ok := asFormattedNumeric(p); ok {
+		return property{Type: t, Coerce: true}
+	}
+
+	typesWithoutNull := func(ts []string) []string {
+		out := []string{}
+		for _, t := range ts {
+			if t != "null" {
+				out = append(out, t)
+			}
+		}
+		return out
+	}
+
+	switch t := typesWithoutNull(p.Inference.Types)[0]; t {
+	case pf.JsonTypeBoolean:
+		return property{Type: elasticTypeBoolean}
+	case pf.JsonTypeInteger:
+		return property{Type: elasticTypeLong}
+	case pf.JsonTypeString:
+		if p.Inference.String_.ContentEncoding == "base64" {
+			return property{Type: elasticTypeBinary}
+		}
+
+		switch f := p.Inference.String_.Format; f {
+		// Formats for "integer" and "number" are handled above.
+		case "date":
+			return property{Type: elasticTypeDate}
+		case "date-time":
+			return property{Type: elasticTypeDate}
+		case "ipv4":
+			return property{Type: elasticTypeIp}
+		case "ipv6":
+			return property{Type: elasticTypeIp}
+		default:
+			if p.IsPrimaryKey {
+				return property{Type: elasticTypeKeyword}
+			} else {
+				return property{Type: elasticTypeText}
+			}
+		}
+	case pf.JsonTypeObject:
+		return property{Type: elasticTypeFlattened}
```

Comment:
Seems fair to `index: false` the root document projection. I'm not sure if it would be desirable to have some amount of indexing to make the field more useful, but it could always be indexed after the fact if that is desirable.